pkg/ring/ring.go

```diff
@@ -326,29 +339,51 @@ func (r *Ring) GetAll(op Operation) (ReplicationSet, error) {
 		return ReplicationSet{}, ErrEmptyRing
 	}
 
+	maxUnavailableZones := len(r.ringZones) / 2
+
 	// Calculate the number of required ingesters;
 	// ensure we always require at least RF-1 when RF=3.
-	numRequired := len(r.ringDesc.Ingesters)
-	if numRequired < r.cfg.ReplicationFactor {
-		numRequired = r.cfg.ReplicationFactor
-	}
-	maxUnavailable := r.cfg.ReplicationFactor / 2
-	numRequired -= maxUnavailable
+	numRequired := calculateRequiredInstances(len(r.ringDesc.Ingesters), r.cfg.ReplicationFactor)
 
-	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	instances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	zoneFailures := make(map[string]int)
 	for _, ingester := range r.ringDesc.Ingesters {
 		if r.IsHealthy(&ingester, op) {
-			ingesters = append(ingesters, ingester)
+			instances = append(instances, ingester)
+
+		} else {
+			zoneFailures[ingester.Zone]++
+		}
+	}
+	maxErrors := len(instances) - numRequired
+
+	if r.cfg.ZoneAwarenessEnabled {
```

Comment:
This would simplify the logic inside the if.

```suggestion
	if r.cfg.ZoneAwarenessEnabled && if len(zoneFailures) > 0 {
```

Reply:
Good catch. The if statement has been updated and the redundant if statement inside has been removed.